File: 01-Languages/Kotlin/22-Mini-Projects/TaskTracker/tests/TaskRepositoryTest.kt
import java.sql.Connection
import java.sql.DriverManager
import kotlin.test.Test
import kotlin.test.BeforeTest
import kotlin.test.AfterTest
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.assertFailsWith

class TaskRepositoryTest {
    private lateinit var conn: Connection
    private lateinit var repo: TaskRepository

    @BeforeTest
    fun setUp() {
        // A FRESH in-memory connection per test, never the real tasks.db file -- SQLite's
        // ":memory:" database only exists for the lifetime of the single connection that
        // created it, so sharing one connection across tests would leak state between them
        // (the exact same reasoning as the C#/Java courses' mini-project test suites).
        Class.forName("org.sqlite.JDBC")
        conn = DriverManager.getConnection("jdbc:sqlite::memory:")
        repo = TaskRepository(conn)
    }

    @AfterTest
    fun tearDown() {
        conn.close()
    }

    @Test
    fun addTaskAssignsIncrementingIds() {
        val first = repo.addTask("Write README", Priority.HIGH)
        val second = repo.addTask("Review PR", Priority.MEDIUM)
        assertEquals(1L, first.id)
        assertEquals(2L, second.id)
    }

    @Test
    fun addedTaskStartsPending() {
        val task = repo.addTask("Water plants", Priority.LOW)
        assertEquals(Status.PENDING, task.status)
    }

    @Test
    fun addTaskRejectsBlankTitle() {
        val exception = assertFailsWith<IllegalArgumentException> {
            repo.addTask("   ", Priority.LOW)
        }
        assertTrue(exception.message!!.contains("blank"))
    }

    @Test
    fun listTasksReturnsAllInInsertionOrder() {
        repo.addTask("First", Priority.LOW)
        repo.addTask("Second", Priority.LOW)
        repo.addTask("Third", Priority.LOW)
        val titles = repo.listTasks().map { it.title }
        assertEquals(listOf("First", "Second", "Third"), titles)
    }

    @Test
    fun listTasksFiltersByStatus() {
        val t1 = repo.addTask("Task A", Priority.LOW)
        repo.addTask("Task B", Priority.LOW)
        repo.markDone(t1.id)

        val pending = repo.listTasks(Status.PENDING)
        val done = repo.listTasks(Status.DONE)

        assertEquals(1, pending.size)
        assertEquals("Task B", pending.single().title)
        assertEquals(1, done.size)
        assertEquals("Task A", done.single().title)
    }

    @Test
    fun markDoneChangesStatus() {
        val task = repo.addTask("Deploy", Priority.HIGH)
        repo.markDone(task.id)
        val reloaded = repo.getTask(task.id)
        assertEquals(Status.DONE, reloaded.status)
    }

    @Test
    fun markDoneOnMissingIdThrows() {
        val exception = assertFailsWith<TaskNotFoundException> {
            repo.markDone(999L)
        }
        assertEquals("No task found with id 999", exception.message)
    }

    @Test
    fun deleteTaskRemovesIt() {
        val task = repo.addTask("Temporary", Priority.LOW)
        repo.deleteTask(task.id)
        assertFailsWith<TaskNotFoundException> { repo.getTask(task.id) }
    }

    @Test
    fun deleteTaskOnMissingIdThrows() {
        assertFailsWith<TaskNotFoundException> { repo.deleteTask(42L) }
    }

    @Test
    fun getStatsCountsPendingAndDone() {
        val t1 = repo.addTask("A", Priority.LOW)
        val t2 = repo.addTask("B", Priority.LOW)
        repo.addTask("C", Priority.LOW)
        repo.markDone(t1.id)
        repo.markDone(t2.id)

        val stats = repo.getStats()
        assertEquals(1, stats.pending)
        assertEquals(2, stats.done)
        assertEquals(3, stats.total)
    }
}

File: 01-Languages/PHP/22-Mini-Projects/TaskTracker/tests/TaskRepositoryTest.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/../src/TaskItem.php';
require_once __DIR__ . '/../src/TaskNotFoundException.php';
require_once __DIR__ . '/../src/TaskRepository.php';

use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\TestCase;

final class TaskRepositoryTest extends TestCase
{
    private TaskRepository $repo;

    protected function setUp(): void
    {
        // A fresh :memory: connection per test -- SQLite's in-memory database
        // only exists for the lifetime of the connection that opened it, so
        // building a new PDO here (rather than sharing one across tests)
        // guarantees no state leaks between test methods, exactly the
        // discipline Lesson 18's own setUp() establishes.
        $this->repo = new TaskRepository(new PDO('sqlite::memory:'));
    }

    #[Test]
    public function addingATaskDefaultsToPendingAndMediumPriority(): void
    {
        $task = $this->repo->add('Write tests');

        $this->assertSame('Write tests', $task->title);
        $this->assertSame(Priority::Medium, $task->priority);
        $this->assertSame(Status::Pending, $task->status);
        $this->assertSame(1, $task->id);
    }

    #[Test]
    public function addingATaskWithAnEmptyTitleThrows(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Task title must not be empty.');

        $this->repo->add('   ');
    }

    #[Test]
    #[DataProvider('priorityCases')]
    public function addingATaskAcceptsEachPriority(Priority $priority, string $expectedLabel): void
    {
        $task = $this->repo->add('Priority check', $priority);

        $this->assertSame($priority, $task->priority);
        $this->assertSame($expectedLabel, $task->priority->label());
    }

    public static function priorityCases(): array
    {
        return [
            'low priority' => [Priority::Low, 'Low'],
            'medium priority' => [Priority::Medium, 'Medium'],
            'high priority' => [Priority::High, 'High'],
        ];
    }

    #[Test]
    public function allReturnsTasksInInsertionOrder(): void
    {
        $this->repo->add('First');
        $this->repo->add('Second');
        $this->repo->add('Third');

        $titles = array_map(fn (TaskItem $t) => $t->title, $this->repo->all());

        $this->assertSame(['First', 'Second', 'Third'], $titles);
    }

    #[Test]
    public function allCanFilterByStatus(): void
    {
        $first = $this->repo->add('Pending one');
        $second = $this->repo->add('Will be done');
        $this->repo->markDone($second->id);

        $pending = $this->repo->all(Status::Pending);
        $done = $this->repo->all(Status::Done);

        $this->assertCount(1, $pending);
        $this->assertSame($first->id, $pending[0]->id);
        $this->assertCount(1, $done);
        $this->assertSame($second->id, $done[0]->id);
    }

    #[Test]
    public function findReturnsTheMatchingTask(): void
    {
        $created = $this->repo->add('Findable');

        $found = $this->repo->find($created->id);

        $this->assertSame($created->title, $found->title);
    }

    #[Test]
    public function findingAMissingIdThrowsTaskNotFoundException(): void
    {
        $this->expectException(TaskNotFoundException::class);
        $this->expectExceptionMessage('No task found with id 999.');

        $this->repo->find(999);
    }

    #[Test]
    public function markDoneFlipsStatusAndReturnsTheUpdatedTask(): void
    {
        $task = $this->repo->add('Flip me');

        $updated = $this->repo->markDone($task->id);

        $this->assertSame(Status::Done, $updated->status);
        // The original TaskItem instance is untouched -- TaskItem is
        // immutable, so markDone() had to hand back a NEW object rather
        // than mutating the one already held by the caller.
        $this->assertSame(Status::Pending, $task->status);
    }

    #[Test]
    public function markDoneOnAMissingIdThrowsTaskNotFoundException(): void
    {
        $this->expectException(TaskNotFoundException::class);

        $this->repo->markDone(42);
    }

    #[Test]
    public function deleteRemovesTheTask(): void
    {
        $task = $this->repo->add('Delete me');

        $this->repo->delete($task->id);

        $this->expectException(TaskNotFoundException::class);
        $this->repo->find($task->id);
    }

    #[Test]
    public function deleteOnAMissingIdThrowsTaskNotFoundException(): void
    {
        $this->expectException(TaskNotFoundException::class);

        $this->repo->delete(7);
    }

    #[Test]
    public function statsCountsPendingAndDoneSeparately(): void
    {
        $a = $this->repo->add('A');
        $this->repo->add('B');
        $this->repo->markDone($a->id);

        $stats = $this->repo->stats();

        $this->assertSame(1, $stats->pending);
        $this->assertSame(1, $stats->done);
        $this->assertSame(2, $stats->total());
    }

    #[Test]
    public function statsOnAnEmptyRepositoryIsAllZero(): void
    {
        $stats = $this->repo->stats();

        $this->assertSame(0, $stats->pending);
        $this->assertSame(0, $stats->done);
        $this->assertSame(0, $stats->total());
    }
}
